### PR TITLE
[DPE-3314] remove secrets on broken

### PR DESCRIPTION
**Pull request by MiaAltieri**
## FYI
most of the work here is done on the `mongos` side, but some changes are needed in the libs that are owned by this charm

## Issue
Secretes associated with username, password, uri are still present on the `host-application` and `mongos sub-charm` relation after relation with `config-server` and `mongos sub-charm` is broken

## Solution
Call appropriate functions from mongos charm

## Future PRs
1. Implementation + Integration tests added on mongos side

### CODE REVIEW

File: lib/charms/mongodb/v1/shards_interface.py [hunk 1]
```diff
@@ -146,6 +146,18 @@ def pass_hook_checks(self, event: EventBase) -> bool:
             event.defer()
             return False
 
+        if isinstance(event, RelationBrokenEvent):
```
Comment (MiaAltieri) [inline lib/charms/mongodb/v1/shards_interface.py:149]: move this due to comments on previous PR

File: src/charm.py [hunk 2]
```diff
@@ -1334,16 +1334,16 @@ def proceed_on_broken_event(self, event) -> int:
                 "Deferring, must wait for relation departed hook to decide if relation should be removed."
             )
             event.defer()
-            return
+            return False
```
Comment (dmitry-ratushnyy) [inline src/charm.py:1335]: Could you please update the function return type as well (not it is "int")

File: lib/charms/mongodb/v1/shards_interface.py [hunk 3]
```diff
@@ -159,9 +171,7 @@ def _on_relation_event(self, event):
 
         departed_relation_id = None
         if isinstance(event, RelationBrokenEvent):
-            departed_relation_id = self.charm.proceed_on_broken_event(event)
-            if not departed_relation_id:
-                return
+            departed_relation_id = event.relation.id
```
Comment (dmitry-ratushnyy) [inline lib/charms/mongodb/v1/shards_interface.py:173]: Looks like here if the event is NOT  `RelationBrokenEvent`  then `departed_relation_id` will be None, so maybe in that case we need to return from function earlier?

File: lib/charms/mongodb/v0/config_server_interface.py [hunk 4]
```diff
@@ -228,6 +233,31 @@ def _on_relation_changed(self, event) -> None:
 
         self.charm.unit.status = ActiveStatus()
 
+    def _on_relation_broken(self, event) -> None:
```
Comment (Mehdi-Bendriss) [inline lib/charms/mongodb/v0/config_server_interface.py]: nit: typed args

File: lib/charms/mongodb/v0/config_server_interface.py [hunk 5]
```diff
@@ -228,6 +233,31 @@ def _on_relation_changed(self, event) -> None:
...
+        # Only relation_deparated events can check if scaling down
+        departed_relation_id = event.relation.id
```
Comment (Mehdi-Bendriss) [inline lib/charms/mongodb/v0/config_server_interface.py]: this variable seems to be only used once on the subsequent condition - safe to drop and refer to `event.relation.id`

File: lib/charms/mongodb/v1/shards_interface.py [hunk 6]
```diff
@@ -146,6 +146,18 @@ def pass_hook_checks(self, event: EventBase) -> bool:
             event.defer()
             return False
 
+        if isinstance(event, RelationBrokenEvent):
+            departed_relation_id = event.relation.id
```
Comment (Mehdi-Bendriss) [inline lib/charms/mongodb/v1/shards_interface.py]: same as previous comment

**Review by Mehdi-Bendriss — COMMENTED** [review]
Good work! I only have a few minor comments

File: lib/charms/mongodb/v1/shards_interface.py (hunk 3 above)
Comment (MiaAltieri) [reply lib/charms/mongodb/v1/shards_interface.py:173]: > Looks like here if the event is NOT RelationBrokenEvent then departed_relation_id will be None

That is correct. I am not 100% sure what you mean by:
> in that case we need to return from function earlier

But, we should **not** return if `departed_relation_id` is `None`

In this function we pass `departed_relation_id` to subsequent functions and those functions handle the cases in which `departed_relation_id` is `None`, namely:
```
            self.add_shards(departed_relation_id)
            self.remove_shards(departed_relation_id)
 ```
 
 does that answer your question? Or am I misunderstanding?

Comment (dmitry-ratushnyy) [reply lib/charms/mongodb/v1/shards_interface.py:173]: Why would we want to do his call ?

```
self.add_shards(None)
self.remove_shards(None)
```

Comment (MiaAltieri) [reply lib/charms/mongodb/v1/shards_interface.py:173]: Please read the functions `add_shards` and `remove_shards` 

ideally from those function signatures it should be obvious why we would want to do this. If it isn't obvious then there is a big problem in the readability of this file and that needs to be updated ASAP

File: src/charm.py [hunk 7]
```diff
@@ -1323,7 +1323,7 @@ def set_scaling_down(self, event: RelationDepartedEvent) -> bool:
         self.unit_peer_data[rel_departed_key] = json.dumps(scaling_down)
         return scaling_down
 
-    def proceed_on_broken_event(self, event) -> int:
+    def proceed_on_broken_event(self, event) -> None:
```
Comment (dmitry-ratushnyy) [inline src/charm.py]: Maybe it should be bool? 🤔

**Review by dmitry-ratushnyy — APPROVED** [review]

**Review by Mehdi-Bendriss — APPROVED** [review]

**Pull request merged**